cro-clib/src/types.rs

```diff
@@ -46,7 +64,50 @@ impl CroResult {
     }
 }
 
+/// copy from c array to rust array
+/// # Safety
+pub unsafe fn copy_string(src: &str, dst: &mut [u8]) {
+    dst[..src.len()].copy_from_slice(&src.as_bytes()[..src.len()]);
+    dst[src.len()] = 0;
+}
+
 /// # Safety
 pub unsafe fn get_string(src: *const c_char) -> String {
     CStr::from_ptr(src).to_string_lossy().into_owned()
 }
+
+#[repr(C)]
+pub struct CroUtxo {
+    pub address: [u8; 32], // holder
+    pub value: u64,        // amount in carson unit
+    pub valid_from: u64,   // time lock
+}
+
+#[derive(Clone, Default)]
+pub struct CroTx {
+    // out amount - in amount = fee
+    // in
+    // if signed, these 3 counts are the same
+    pub txin_pointer: Vec<TxoPointer>,          // which utxo?
+    pub txin_witness: Vec<Option<TxInWitness>>, // sinature
+    pub txin: Vec<TxOut>,                       // utxo which will be consumed
+    // out
+    pub txout: Vec<TxOut>, // new utxo
+    // viewkeys which can view thix tx
+    pub viewkey: Vec<secp256k1::PublicKey>,
```

Comment:
this structure duplicates https://github.com/crypto-com/chain/blob/master/chain-core/src/tx/mod.rs#L60
As the internal structure of `CroTx` is oblivious to C, this could just contain `Tx` and `TxWitness` and mutate these (`(something).tx.add_input(..)`, `(something).tx.add_output(..)`: https://github.com/crypto-com/chain/blob/master/chain-core/src/tx/data/mod.rs#L111 etc.)

Reply:
ok, i'll try to use it directly.